### DISCUSSION TITLE
Guid: Literal or implicit cast to and from string

### DISCUSSION THREAD

**Discussion by MBAatEJOT** [dotnet/csharplang#5308] [Language Ideas]
To create a Guid these days - as far as I know - you have to use either one of the following:
```csharp
guid = new Guid();
guid = new Guid("0faa9318-6408-47ca-b0ca-28c7482bdb09");
guid = Guid.NewGuid();
guid = Guid.Parse("0faa9318-6408-47ca-b0ca-28c7482bdb09");
// and of course Guid.ParseExact, Guid.TryParse, Guid.TryParseExact
```
`new Guid();` and `Guid.NewGuid();` are fine for creating a Guid where you don't have to know the value at coding time. But if you need to know a guid value you've to use one of the other variants. For a single value it's fine as well but if you work with the entity framework and you have to seed some data it's to much code - in my opinion:

```csharp
entity.HasData(
	new { Id = new Guid("6c6502a2-0a78-4f3a-a28e-4d3a0835719a"), Property1 = "valuesforprop1", Property2 = "othervalues", AndSoOn = 123 },
	new { Id = new Guid("77a47113-55e4-4697-95ac-57bdb5075be2"), Property1 = "valuesforprop1", Property2 = "othervalues", AndSoOn = 123 },
	new { Id = new Guid("97f47c52-e57d-42a9-9627-552642ccb30e"), Property1 = "valuesforprop1", Property2 = "othervalues", AndSoOn = 123 },
	new { Id = new Guid("6ccb26c4-1c4c-41ae-99a4-982f08817f6c"), Property1 = "valuesforprop1", Property2 = "othervalues", AndSoOn = 123 },
	new { Id = new Guid("e8555568-9994-47a2-a26e-22a756d020c7"), Property1 = "valuesforprop1", Property2 = "othervalues", AndSoOn = 123 },
	new { Id = new Guid("44f9c50d-1fdd-4bfc-94f9-ff9f8f26c02e"), Property1 = "valuesforprop1", Property2 = "othervalues", AndSoOn = 123 },
	new { Id = new Guid("a136d4ba-f0a9-4855-ab74-7eeac629d6de"), Property1 = "valuesforprop1", Property2 = "othervalues", AndSoOn = 123 },
	new { Id = new Guid("58cd4461-cbf6-4443-86a8-cd68e1448ecb"), Property1 = "valuesforprop1", Property2 = "othervalues", AndSoOn = 123 },
	new { Id = new Guid("b9692f23-878c-4f3f-ac55-912f8fa243bb"), Property1 = "valuesforprop1", Property2 = "othervalues", AndSoOn = 123 },
	new { Id = new Guid("8b3d20f1-f1da-4e1d-af46-7297bc83e3a0"), Property1 = "valuesforprop1", Property2 = "othervalues", AndSoOn = 123 },
	new { Id = new Guid("84f36a31-9925-4f6a-92f8-ba23e36c8001"), Property1 = "valuesforprop1", Property2 = "othervalues", AndSoOn = 123 },
	// ... more Data
);
```

It would be easier and much less overhead if there would be either some kind of Guid literal like 
```csharp
Guid id = "84f36a31-9925-4f6a-92f8-ba23e36c8001"G // <- like for numeric literals
```

or a implicit type conversion

```csharp
Guid id = "84f36a31-9925-4f6a-92f8-ba23e36c8001";
string strId = id;
```

**Comment by svick** [comment]
You can use target-typed `new` to shorten your code: `new("84f36a31-9925-4f6a-92f8-ba23e36c8001")`.

**Reply by MBAatEJOT** [reply]
True, that's shorter but it still leads to 4 - 5 characters more code to write. It's still not optimal.

**Reply by amis92** [reply]
Create a local function "CreateRow" or sth with all necessary arguments. You won't need to repeat property names, and can target-type everything you need. Profit!

Adding implicit casts just because you want to save a couple characters doesn't seem like a valid request. Additionally, it's a runtime (BCL) request, not a C# language one. :)

**Reply by MBAatEJOT** [reply]
> Create a local function "CreateRow" or sth with all necessary arguments. You won't need to repeat property names, and can target-type everything you need. Profit!

Wouldn't work for seeding migration data because the ids have always be the same, using a local function would mean, that every time, a migration would be added, the ids would change. This would be really, really bad.

> Adding implicit casts just because you want to save a couple characters doesn't seem like a valid request. Additionally, it's a runtime (BCL) request, not a C# language one. :)

Yeah, but that's just an alternative. My first suggestion is using a literal, which in my understanding is a request for the C# language

> Adding implicit casts just because you want to save a couple characters 
Everything is relative. Only one or two lines where you have to use a guid might be called a couple of characters. Seeding for example 100 lines of code would save 400 - 500 characters of code. If you have to use foreign keys which are also Guids would save 800 - 1000 characters. It depends how often you use the literal / implicit conversion.

Nonetheless it would make the code more readable.

**Reply by amis92** [reply]
```csharp

static YourDto CreateDto(string id, string prop1, string prop2, int andSoOn, etc.) => new YourDto
{
    Id = new Guid(id),
    Property1 = prop1,
    Property2 = prop2,
    AndSoOn = andSoOn,
    etc.
};

entity.HasData(
	CreateDto("6c6502a2-0a78-4f3a-a28e-4d3a0835719a", "valuesforprop1", "othervalues", 123),
	CreateDto("77a47113-55e4-4697-95ac-57bdb5075be2", "valuesforprop1", "othervalues", 123),
	CreateDto("97f47c52-e57d-42a9-9627-552642ccb30e", "valuesforprop1", "othervalues", 123),
	CreateDto("e8555568-9994-47a2-a26e-22a756d020c7", "valuesforprop1", "othervalues", 123),
	CreateDto("44f9c50d-1fdd-4bfc-94f9-ff9f8f26c02e", "valuesforprop1", "othervalues", 123),
	CreateDto("a136d4ba-f0a9-4855-ab74-7eeac629d6de", "valuesforprop1", "othervalues", 123),
	CreateDto("58cd4461-cbf6-4443-86a8-cd68e1448ecb", "valuesforprop1", "othervalues", 123),
	CreateDto("b9692f23-878c-4f3f-ac55-912f8fa243bb", "valuesforprop1", "othervalues", 123),
	CreateDto("8b3d20f1-f1da-4e1d-af46-7297bc83e3a0", "valuesforprop1", "othervalues", 123),
	CreateDto("84f36a31-9925-4f6a-92f8-ba23e36c8001", "valuesforprop1", "othervalues", 123),
	// ... more Data
);
```

How does that not keep IDs the same? it's just moving the initializer into a separate method. No magic, you just pass the parameters to your own method instead of writing the initializer 800 times.

I'll leave counting saved characters as an exercise to you :D

**Reply by MBAatEJOT** [reply]
Ok, sorry I missunderstand your suggestion.

That is a suitable workaround. 

But nonetheless what would speak against a literal / an implicit conversion? It would only bring benefits to the language.

Less and more readable code, like when 
```csharp
public string MyProp
{
   get { return _myProp; }
   set { _myProp = value; }
}
```
became

```csharp
public string MyProp
{
   get;
   set;
}
```

**Reply by HaloFour** [reply]
I'd argue that an implicit conversion is not appropriate as the conversion is fallible.

As for making a literal, that makes more sense, but I don't think there'd be much ROI on such a change.  It's incredibly uncommon for someone to have to include a bunch of explicit GUIDs in their code vs. having them generated.  And in those uncommon cases target-typed new already gives you a fairly terse shorthand.  I don't see a lot of utility in trying to make that code shorter.

**Reply by svick** [reply]
@MBAatEJOT There are many ways to make code more succinct, but not all of them are worth it, especially since they make the language more complicated and can make the code less readable.

Auto-properties are very common, avoid repetitive and potentially buggy code and shorten the code fairly significantly, which makes them worthwhile. Your suggestion is none of those things, which makes it much less worthy.

**Reply by MBAatEJOT** [reply]
Ok, auto-properties are pretty common, but what's about the *target-typed new*? It's not that helpful when initializing a variable. Only saves a few characters as well.

```csharp
var obj = new MyObject();
MyObject obj = new ();
```

**Reply by svick** [reply]
@MBAatEJOT Target-typed `new` can save a lot in contexts where you can't use `var`, e.g.:

```c#
Dictionary<string, Func<ParseResult, CommandBase>> SubCommands => new Dictionary<string, Func<ParseResult, CommandBase>> { ... };
Dictionary<string, Func<ParseResult, CommandBase>> SubCommands => new() { ... };
```

And again, it also has the advantage of avoiding repetition.

**Reply by HaloFour** [reply]
The benefit for target-typed new is certainly much greater when the type name is long, especially in the case of generic types.  But more importantly, target-typed new can apply to any type and used in numerous scenarios, whereas a Guid literal could only be applied in one place, and a fairly uncommon place.

If the proposal were to instead have a means to declare a string literal format for an arbitrary struct type I think I could get behind that.  I believe such a proposal would need to rely on both constant expressions and blittable structs, as well as possibly changes to the runtime to support use in attributes and the like.  But at least then it would be a general purpose feature that could be applied to any type and Guid would just happen to able to benefit from it.
